addons/mail/static/src/components/composer/composer_tests.js

```diff
@@ -1141,6 +1141,105 @@ QUnit.test('warning on send with shortcut when attempting to post message with s
+            }
+            return res;
+        }
+    });
+    const composer = this.env.models['mail.composer'].create();
+    await this.createComposerComponent(composer);
+    const file = await createFile({
+        content: 'hello, world',
+        contentType: 'text/plain',
+        name: 'text.txt',
+    });
+    await afterNextRender(() =>
+        inputFiles(
+            document.querySelector('.o_FileUploader_input'),
+            [file]
+        )
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Composer_attachmentList',
+        "should have an attachment list"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Composer .o_Attachment',
+        "should have only one attachment"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Composer .o_Attachment.o-temporary',
+        "should have an uploading attachment"
+    );
+
+    await afterNextRender(() =>
+        document.querySelector('.o_Attachment_asideItemUnlink').click());
+    assert.containsNone(
+        document.body,
+        '.o_Composer .o_Attachment',
+        "should not have any attachment left after unlinking temporary one"
+    );
+});
+
+QUnit.test('remove an uploading attachment aborts upload', async function (assert) {
+    // Test is kind of technical as it's difficult to test that fetch requests
+    // are cancelled without checking attachment model fields.
+    assert.expect(3);
+
+    await this.start({
+        async mockFetch(resource, init) {
+            const res = this._super(...arguments);
+            if (resource === '/web/binary/upload_attachment') {
+                // simulates uploading indefinitely
+                await new Promise(() => {});
+            }
+            return res;
+        }
+    });
+    const composer = this.env.models['mail.composer'].create();
+    await this.createComposerComponent(composer);
+    const file = await createFile({
+        content: 'hello, world',
+        contentType: 'text/plain',
+        name: 'text.txt',
+    });
+    await afterNextRender(() =>
+        inputFiles(
+            document.querySelector('.o_FileUploader_input'),
+            [file]
+        )
+    );
+    const attachmentLocalId = document.querySelectorAll('.o_Attachment')[0].dataset.attachmentLocalId;
+    const attachmentAbortController =
+        this.env.models['mail.attachment'].get(attachmentLocalId).uploadingAbortController;
+    assert.ok(
+        attachmentAbortController,
+        'Uploading attachment should have an abort controller',
+    );
+    assert.notOk(
+        attachmentAbortController.signal.aborted,
+        'Uploading attachment fetch request should not be aborted yet',
+    );
+
+    await afterNextRender(() =>
+        document.querySelector('.o_Attachment_asideItemUnlink').click());
+    assert.ok(
+        attachmentAbortController.signal.aborted,
+        'fetch request should be aborted after uploading attachment has been removed',
+    );
```

Comment:
You could have avoided leaking the abort controller by just asserting the `upload_attachment` fetch has been aborted. It's hardly doable because it's on `fetch` API and also because `AbortError` is silently ignored in code.

Eventually, the code will be in models and we want to have logging of some sort. For the time being, it's acceptable to have trigger in component that are kind of logs. I'm good with
```js
this.trigger('o-attachment-upload-abort', { /* data identifying attachment that has been aborted */ });
```
And catch this in test.
With all this, you don't need model test and you don't have to leak too much technical aspects. Sure it assumes the trigger exists... But that's more robust that current solution, and it's closer that what we want later.

_(sorry for not having suggested this solution at first, and instead let you make a model test for abort controller...)_

The test of field uploadingAbortController can be removed, because this is too technical and all in all it's code that just add too much constraint, so better drop it.

Reply:
How do you catch the event ? with `addEventListener` or `afterEvent` ?

Reply:
`afterEvent` does `addEventListener` and `await` for you, so use `afterEvent`.